presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/FunctionDescriptorParser.java

```diff
@@ -0,0 +1,126 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.scalar.annotations;
+
+import com.facebook.presto.common.Subfield;
+import com.facebook.presto.spi.function.ComplexTypeFunctionDescriptor;
+import com.facebook.presto.spi.function.LambdaArgumentDescriptor;
+import com.facebook.presto.spi.function.LambdaDescriptor;
+import com.facebook.presto.spi.function.ScalarFunctionDescriptor;
+import com.facebook.presto.spi.function.ScalarFunctionLambdaArgumentDescriptor;
+import com.facebook.presto.spi.function.ScalarFunctionLambdaDescriptor;
+import com.facebook.presto.spi.function.StaticMethodPointer;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class FunctionDescriptorParser
+{
+    private FunctionDescriptorParser() {}
+    public static ComplexTypeFunctionDescriptor parseFunctionDescriptor(ScalarFunctionDescriptor descriptor)
+    {
+        if (descriptor.outputToInputTransformationFunction().length > 1) {
+            throw new IllegalArgumentException("outputToInputTransformationFunction must contain at most 1 element.");
+        }
+        return new ComplexTypeFunctionDescriptor(
+                descriptor.isAccessingInputValues(),
+                parseLambdaDescriptors(descriptor.lambdaDescriptors()),
+                descriptor.argumentIndicesContainingMapOrArray().length == 1 ?
+                        Optional.of(ImmutableSet.copyOf(Arrays.stream(descriptor.argumentIndicesContainingMapOrArray()[0].value()).iterator())) : Optional.empty(),
+                descriptor.outputToInputTransformationFunction().length == 1 ?
+                        Optional.of(parseSubfieldTransformationFunction(descriptor.outputToInputTransformationFunction()[0])) :
+                        Optional.empty());
+    }
+
+    private static Function<Set<Subfield>, Set<Subfield>> parseSubfieldTransformationFunction(StaticMethodPointer staticMethodPointer)
+    {
+        Method subfieldTransformationMethod;
+        try {
+            subfieldTransformationMethod = ((Class<?>) staticMethodPointer.clazz()).getDeclaredMethod(staticMethodPointer.method(), Set.class);
+        }
+        catch (NoSuchMethodException e) {
+            return null;
```

Comment:
Better to return Optional here to avoid potential npes in future

Reply:
Instead of returning `null`, will throw  PrestoException with `FUNCTION_IMPLEMENTATION_ERROR` category.